src/allmydata/test/test_storage_https.py

```diff
@@ -0,0 +1,219 @@
+        """
+        private_key = generate_private_key()
+        certificate = generate_certificate(private_key)
+        async with self.listen(
+            private_key_to_file(FilePath(self.mktemp()), private_key),
+            cert_to_file(FilePath(self.mktemp()), certificate),
+        ) as url:
+            response = await self.request(url, certificate)
+            self.assertEqual(await response.content(), b"YOYODYNE")
+
+    @async_to_deferred
+    async def test_server_certificate_has_wrong_hash(self):
+        """
+        If the server's certificate hash doesn't match the hash the client
+        expects, the request to the server fails.
+        """
+        private_key1 = generate_private_key()
+        certificate1 = generate_certificate(private_key1)
+        private_key2 = generate_private_key()
+        certificate2 = generate_certificate(private_key2)
+
+        async with self.listen(
+            private_key_to_file(FilePath(self.mktemp()), private_key1),
+            cert_to_file(FilePath(self.mktemp()), certificate1),
+        ) as url:
+            with self.assertRaises(ResponseNeverReceived):
+                await self.request(url, certificate2)
+
+    @async_to_deferred
+    async def test_server_certificate_expired(self):
+        """
+        If the server's certificate has expired, the request to the server
+        succeeds if the hash matches the one the client expects; expiration has
+        no effect.
+        """
+        private_key = generate_private_key()
+        certificate = generate_certificate(private_key, expires_days=-10)
+
+        async with self.listen(
+            private_key_to_file(FilePath(self.mktemp()), private_key),
+            cert_to_file(FilePath(self.mktemp()), certificate),
+        ) as url:
+            response = await self.request(url, certificate)
+            self.assertEqual(await response.content(), b"YOYODYNE")
+
+    @async_to_deferred
+    async def test_server_certificate_not_valid_yet(self):
+        """
+        If the server's certificate is only valid starting in The Future, the
+        request to the server succeeds if the hash matches the one the client
+        expects; start time has no effect.
+        """
+        private_key = generate_private_key()
+        certificate = generate_certificate(
+            private_key, expires_days=10, valid_in_days=5
+        )
+
+        async with self.listen(
+            private_key_to_file(FilePath(self.mktemp()), private_key),
+            cert_to_file(FilePath(self.mktemp()), certificate),
+        ) as url:
+            response = await self.request(url, certificate)
+            self.assertEqual(await response.content(), b"YOYODYNE")
+
+    # A potential attack to test is a private key that doesn't match the
+    # certificate... but OpenSSL (quite rightly) won't let you listen with that
+    # so I don't know how to test that! See
+    # https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3884
```

Comment:
One other error case that came to mind is the case of a server using a CA-issued certificate.  I think this _mostly_ overlaps with "has wrong hash" which is tested but I guess it might be nice to test separately just to be sure OpenSSL isn't somehow deciding to jump over all of our logic because it sees a signature from a well-known CA.

Of course, how to _get_ a CA-issued certificate to use in the tests is an interesting question.

I don't consider this a blocker for this PR but it might be another case to track on 3884 or in another ticket.